File: src/Day06.hs
module Day06 where

import AOCSolution (getSolution)
import Common
import Data.Bifunctor (second)
import Data.List (transpose)
import Data.List.Split (chunksOf)
import Inputs (InputType (..), readInput)

data Op = Multiply | Add deriving (Show)

day06SampleInput, day06ActualInput :: IO (Maybe String)
day06SampleInput = readInput Sample 6
day06ActualInput = readInput Actual 6

day06 :: IO ((String, String), (String, String))
day06 = do
  s <- maybe ("", "") solve <$> day06SampleInput
  a <- maybe ("", "") solve <$> day06ActualInput
  return (s, a)

-- >>> day06
-- (("4277556","3263827"),("6635273135233","12542543681221"))

solve :: String -> (String, String)
solve = getSolution parseInput part1 part2

parseInput :: String -> [(Op, [String])]
parseInput input = parseOp (init i) (last i ++ " ")
  where
    i = lines input

parseOp :: [String] -> String -> [(Op, [String])]
parseOp _ [] = []
parseOp nums (o : ops) = (op, ns) : parseOp nums' ops'
  where
    n = length $ takeWhile (== ' ') ops
    op = getOp [o]
    ops' = drop n ops
    nums' = map (drop $ succ n) nums
    ns = map (take n) nums

part1 :: [(Op, [String])] -> Int
part1 = solution . map (second f)
  where
    f = map (read . filter (/= ' '))

part2 :: [(Op, [String])] -> Int
part2 = solution . map (second makeNums)
  where
    makeNums :: [String] -> [Int]
    makeNums xs | all (== "") xs = []
    makeNums xs = i : makeNums (map tail xs)
      where
        i = read $ filter (/= ' ') $ map head xs

runOp :: Op -> [Int] -> Int
runOp Multiply = product
runOp Add = sum

solution :: [(Op, [Int])] -> Int
solution = sum . map (uncurry runOp)

getOp :: String -> Op
getOp = \case
  "*" -> Multiply
  "+" -> Add

File: src/day06.rs
use std::fs;

/// Sample
/// ```
/// use advent_of_code_2025::{check_day_output, day06::run};
/// check_day_output!(run, "sample", "4277556", "3263827");
/// ```
///
/// Actual
/// ```
/// use advent_of_code_2025::{check_day_output, day06::run};
/// check_day_output!(run, "actual", "6635273135233", "12542543681221");
/// ```
pub fn run(file_type: &str) -> Option<(String, String)> {
    if let Ok(contents) = fs::read_to_string(format!("inputs/{}/06.txt", file_type)) {
        let input = parse_input(&contents);

        Some((part1(&input).to_string(), part2(&input).to_string()))
    } else {
        None
    }
}

fn parse_input(input: &str) -> Vec<(Vec<String>, char)> {
    let lines: Vec<&str> = input.lines().collect();
    let nums = &lines[..lines.len() - 1];
    let ops = lines[lines.len() - 1]
        .chars()
        .chain(" @".chars())
        .collect::<Vec<char>>();

    let mut res = Vec::new();
    let mut i = 0;

    for (j, op) in ops[1..].iter().enumerate() {
        if *op != ' ' {
            res.push((
                nums.iter()
                    .map(|line| line[i..j].to_string())
                    .collect::<Vec<String>>(),
                ops[i],
            ));
            i = j + 1;
        }
    }

    res
}

fn solve<F, I>(input: &[(Vec<String>, char)], transform: F) -> u128
where
    F: Fn(&[String]) -> I,
    I: IntoIterator<Item = String>,
{
    input.iter().fold(0, |acc, (nums, op)| {
        let values: Vec<u128> = transform(nums)
            .into_iter()
            .map(|num| num.trim().parse::<u128>().unwrap())
            .collect();

        acc + match op {
            '+' => values.iter().sum::<u128>(),
            '*' => values.iter().product::<u128>(),
            _ => panic!("Unknown operator"),
        }
    })
}

fn part1(input: &[(Vec<String>, char)]) -> u128 {
    solve(input, |nums| nums.to_vec())
}

fn part2(input: &[(Vec<String>, char)]) -> u128 {
    solve(input, transpose_strings)
}

fn transpose_strings(v: &[String]) -> Vec<String> {
    if v.is_empty() {
        return vec![];
    }

    let rows = v.len();
    let cols = v[0].len();

    let bytes: Vec<&[u8]> = v.iter().map(|s| s.as_bytes()).collect();

    (0..cols)
        .map(|i| {
            let mut s = String::with_capacity(rows);
            for row in &bytes {
                s.push(row[i] as char);
            }
            s
        })
        .collect()
}
